### ISSUE TITLE
compiletest should recover messages from json even if primary span is a macro

### ISSUE THREAD

**Issue by cengiz-io** [rust-lang/rust#40017]
Hello!

This is something that me and @nikomatsakis encountered while trying to write a compile-fail test with a macro as a primary span.

Assume you have a `compile-fail` test like following:

```rust
fn main() {
    let x = vec![]; //~ ERROR 0282
}
```

When you run the test, you get an explicit panic

```
Check compiletest suite=compile-fail mode=compile-fail (x86_64-apple-darwin -> x86_64-apple-darwin)

running 1 test
test [compile-fail] compile-fail/issue-99999.rs ... FAILED

failures:

---- [compile-fail] compile-fail/issue-99999.rs stdout ----

error: /Users/cengizIO/Code/rust-src/src/test/compile-fail/issue-99999.rs:2: expected error not found: E0282

error: 0 unexpected errors found, 1 expected errors not found
status: exit code: 101
command: /Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/stage2/bin/rustc /Users/cengizIO/Code/rust-src/src/test/compile-fail/issue-99999.rs -L /Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/test/compile-fail --target=x86_64-apple-darwin --error-format json -L /Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/test/compile-fail/issue-99999.stage2-x86_64-apple-darwin.compile-fail.libaux -C prefer-dynamic -o /Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/test/compile-fail/issue-99999.stage2-x86_64-apple-darwin -Crpath -O -Lnative=/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/native/rust-test-helpers
not found errors (from test file): [
    Error {
        line_num: 2,
        kind: Some(
            Error
        ),
        msg: "E0282"
    }
]

thread '[compile-fail] compile-fail/issue-99999.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:1122
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    [compile-fail] compile-fail/issue-99999.rs

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured

thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:330


command did not execute successfully: "/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/stage0-tools/x86_64-apple-darwin/release/compiletest" "--compile-lib-path" "/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/stage2/lib" "--run-lib-path" "/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/stage2/lib/rustlib/x86_64-apple-darwin/lib" "--rustc-path" "/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/stage2/bin/rustc" "--rustdoc-path" "/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/stage2/bin/rustdoc" "--src-base" "/Users/cengizIO/Code/rust-src/src/test/compile-fail" "--build-base" "/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/test/compile-fail" "--stage-id" "stage2-x86_64-apple-darwin" "--mode" "compile-fail" "--target" "x86_64-apple-darwin" "--host" "x86_64-apple-darwin" "--llvm-filecheck" "/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/llvm/build/bin/FileCheck" "--nodejs" "/usr/local/bin/node" "--host-rustcflags" "-Crpath -O" "--target-rustcflags" "-Crpath -O -Lnative=/Users/cengizIO/Code/rust-src/build/x86_64-apple-darwin/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python" "--lldb-version" "lldb-360.1.70" "--lldb-python-dir" "/Library/Developer/CommandLineTools/Library/PrivateFrameworks/LLDB.framework/Resources/Python" "--llvm-version" "3.9.1\n" "issue-99999" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp" "--android-cross-path" ""
expected success, got: exit code: 101
```

But, that `issue-99999.rs` indeed fails to compile with an `E0282`.

```
error[E0282]: type annotations needed
 --> src/test/compile-fail/issue-99999.rs:2:13
  |
2 |     let x = vec![]; //~ ERROR E0282
  |         -   ^^^^^^ cannot infer type for `T`
  |         |
  |         consider giving `x` a type
  |
  = note: this error originates in a macro outside of the current crate

error: aborting due to previous error
```

So there must be something wrong with the way that spans are parsed.

@nikomatsakis did a temporary fix for this by changing [src/tools/compiletest/src/json.rs](https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/json.rs) but implementation was not perfect and we ditched it. 

I'm not sure who to cc with this, but thanks in advance!

**Comment by keeperofdakeys** [comment]
What would you expect to happen?

**Comment by cengizIO** [comment]
Adding this patch as a starting point: https://gist.github.com/cengizIO/515a275123d8fe43bb92692bf68601fd

**Comment by nikomatsakis** [comment]
The problem, iirc, was that the *actual span* where the error occurred was outside the current crate, but there were lines from the current crate somewhere in the stack backtrace. I think my preference in general is to remove compile-fail and move to ui testing in general, which would make this a non-issue.

**Comment by Enselic** [comment]
> I think my preference in general is to remove compile-fail and move to ui testing in general, which would make this a non-issue.

Triage: Now that compile-fail is in practice removed, let's close this.
